src/mlpack/methods/ann/rbm/rbm.hpp

```diff
@@ -0,0 +1,447 @@
+   * @param output Hidden neuron activations.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, BinaryRBM>::value, void>::type
+  HiddenMean(DataType&& input, DataType&& output);
+
+  /**
+   * The function calculates the mean of the Normal distribution of P(s|v, h).
+   * The mean is given by:
+   * $h_i*\alpha^{-1}*W_i^T*v$
+   * The variance is given by:
+   * $\alpha^{-1}$
+   * 
+   * @param input Visible layer neurons.
+   * @param output Consists of both the spike samples and slab samples.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  HiddenMean(DataType&& input, DataType&& output);
+
+  /**
+   * The function calculates the mean of the distribution P(h|v),
+   * where mean is given by:
+   * $sigm(v^T*W_i*\alpha_i^{-1}*W_i^T*v + b_i)$
+   *
+   * @param visible The visible layer neurons.
+   * @param spikeMean Indicates P(h|v).
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SpikeMean(DataType&& visible, DataType&& spikeMean);
+
+  /**
+   * The function samples the spike function using Bernoulli distribution.
+   * @param spikeMean Indicates P(h|v).
+   * @param spike Sampled binary spike variables.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SampleSpike(DataType&& spikeMean, DataType&& spike);
+
+  /**
+   * The function calculates the mean of Normal distribution of P(s|v, h),
+   * where the mean is given by:
+   * $h_i*\alpha^{-1}*W_i^T*v$
+   * 
+   * @param visible The visible layer neurons.
+   * @param spike The spike variables from hidden layer.
+   * @param slabMean The mean of the Normal distribution of slab neurons.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SlabMean(DataType&& visible, DataType&& spike, DataType&& slabMean);
+
+  /**
+   * The function samples from the Normal distribution of P(s|v, h),
+   * where the mean is given by:
+   * $h_i*\alpha^{-1}*W_i^T*v$
+   * and variance is given by:
+   * $\alpha^{-1}$
+   *
+   * @param slabMean Mean of the Normal distribution of the slab neurons.
+   * @param slab Sampled slab variable from the Normal distribution.
+   */
+  template<typename Policy = PolicyType>
+  typename std::enable_if<std::is_same<Policy, SpikeSlabRBM>::value, void>::type
+  SampleSlab(DataType&& slabMean, DataType&& slab);
+
+  /**
+   * This function does the k-step Gibbs Sampling.
+   *
+   * @param input Input to the Gibbs function.
+   * @param output Used for storing the negative sample.
+   * @param steps Number of Gibbs Sampling steps taken.
+   */
+  void Gibbs(arma::Mat<ElemType>&& input,
+             arma::Mat<ElemType>&& output,
+             size_t steps = SIZE_MAX);
```

Comment:
To be consistent with the rest of the codebase let's use ``const size_t steps`` here.